Implement most of CSSMathClamp

achristensen07 opened the pull request:
#### 9702a9c388b409636ec9d4ccc08558558037973a
<pre>
Implement most of CSSMathClamp
<a href="https://bugs.webkit.org/show_bug.cgi?id=243956">https://bugs.webkit.org/show_bug.cgi?id=243956</a>

Reviewed by Chris Dumez.

Also update the WPT tests that cover it and the other parts of the css-typed-om.

* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/idlharness-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/perspective-typed-arithmetic-crash.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/resources/testhelper.js:
(assert_style_value_equals):
(createDivWithoutStyle):
(assert_matrix_approx_equals):
(remove_leading_spaces): Deleted.
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/set-css-wide-in-custom-property-crash.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-normalization/normalize-tokens.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-normalization/normalize-tokens.tentative.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-normalization/transformvalue-normalization.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-normalization/transformvalue-normalization.tentative.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-serialization/crashtests/cssInvertValue-convert-crash.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-serialization/crashtests/cssInvertValue-zero.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-serialization/cssMathValue.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-serialization/cssMathValue.tentative.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-serialization/cssTransformValue.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-serialization/cssTransformValue.tentative.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-subclasses/cssColorValue-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-subclasses/cssColorValue.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-subclasses/cssHWB-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-subclasses/cssHWB.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-subclasses/cssPerspective.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-subclasses/cssPerspective.tentative.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-subclasses/numeric-objects/cssMathValue.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-subclasses/numeric-objects/cssMathValue.tentative.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-subclasses/numeric-objects/equals.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-subclasses/numeric-objects/equals.tentative.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-subclasses/numeric-objects/parse.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-subclasses/numeric-objects/parse.tentative.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-subclasses/numeric-objects/to.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-subclasses/numeric-objects/to.tentative.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/computed/getAll-disconnected-element-crash.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/inline/get-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/inline/get-shorthand-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/inline/get-shorthand.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/inline/get.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/contain.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/container-name-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/container-name.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/container-type-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/container-type.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/font-synthesis.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/overflow-clip-margin-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/overflow-clip-margin.html: Copied from LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/font-synthesis.html.
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/width-by-clamp-px-em-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/width-by-clamp-px-em.html: Added.
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/JSCSSStyleValueCustom.cpp:
(WebCore::toJSNewlyCreated):
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/css/typedom/CSSStyleValue.h:
(WebCore::isCSSNumericValue):
(WebCore::isCSSMathValue):
* Source/WebCore/css/typedom/numeric/CSSMathClamp.cpp: Added.
(WebCore::CSSMathClamp::create):
(WebCore::CSSMathClamp::CSSMathClamp):
(WebCore::CSSMathClamp::serialize const):
(WebCore::CSSMathClamp::toSumValue const):
(WebCore::CSSMathClamp::equals const):
* Source/WebCore/css/typedom/numeric/CSSMathClamp.h: Added.
(isType):
* Source/WebCore/css/typedom/numeric/CSSMathClamp.idl: Copied from Source/WebCore/css/typedom/numeric/CSSMathOperator.idl.
* Source/WebCore/css/typedom/numeric/CSSMathInvert.cpp:
(WebCore::CSSMathInvert::equals const):
* Source/WebCore/css/typedom/numeric/CSSMathOperator.h:
(): Deleted.
* Source/WebCore/css/typedom/numeric/CSSMathOperator.idl:

Canonical link: <a href="https://commits.webkit.org/253604@main">https://commits.webkit.org/253604@main</a>
</pre>

















<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4b09ece4ae4c268cb9256f6654e2d5075917b15d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/86510 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/30437 "Built successfully") | [✅ ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/17471 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/95358 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/149071 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/28860 "Built successfully") | [✅ ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/25401 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/78673 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/90602 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/92127 "Passed tests") | [✅ ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/68/builds/23389 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/73461 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/23452 "Passed tests") | 
| | [✅ ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/78398 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/17471 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/66457 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/66/builds/26732 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/17471 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/26646 "Built successfully") | [✅ ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/17471 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2567 "Built successfully and passed tests") | [✅ ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/63/builds/28325 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/73461 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/28266 "Built successfully") | [✅ ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/17471 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-normalization/normalize-tokens.tentative.html:
```diff
@@ -57,6 +57,7 @@
 
   test(t => {
     assert_string_normalizes_to(t, 'margin', value, new CSSUnparsedValue(expectedResult));
+    assert_false(true);
```
uh what?

achristensen07 replied:
This is in the wpt repo.

cdumez commented on LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-normalization/normalize-tokens.tentative.html:
```diff
@@ -57,6 +57,7 @@
 
   test(t => {
     assert_string_normalizes_to(t, 'margin', value, new CSSUnparsedValue(expectedResult));
+    assert_false(true);
```
This seems wrong and we should probably fix the test and merge upstream.

cdumez commented on Source/WebCore/css/typedom/numeric/CSSMathClamp.cpp:
```diff
@@ -0,0 +1,96 @@
+{
+    // https://drafts.css-houdini.org/css-typed-om/#calc-serialization
+    OptionSet<SerializationArguments> serializationArguments { SerializationArguments::Nested, SerializationArguments::WithoutParentheses };
+    builder.append("clamp(");
```
"clamp("_s

cdumez commented on Source/WebCore/css/typedom/numeric/CSSMathClamp.cpp:
```diff
@@ -0,0 +1,96 @@
+    OptionSet<SerializationArguments> serializationArguments { SerializationArguments::Nested, SerializationArguments::WithoutParentheses };
+    builder.append("clamp(");
+    m_lower->serialize(builder, serializationArguments);
+    builder.append(", ");
```
ditto.

cdumez commented on Source/WebCore/css/typedom/numeric/CSSMathClamp.cpp:
```diff
@@ -0,0 +1,96 @@
+    m_lower->serialize(builder, serializationArguments);
+    builder.append(", ");
+    m_value->serialize(builder, serializationArguments);
+    builder.append(", ");
```
ditto.

cdumez commented on Source/WebCore/css/typedom/numeric/CSSMathClamp.h:
```diff
@@ -0,0 +1,63 @@
+    WTF_MAKE_ISO_ALLOCATED(CSSMathClamp);
+public:
+    static ExceptionOr<Ref<CSSMathClamp>> create(CSSNumberish&&, CSSNumberish&&, CSSNumberish&&);
+    CSSNumericValue& lower() { return m_lower.get(); }
```
Do those getters really need to be non-const?

cdumez commented on Source/WebCore/css/typedom/numeric/CSSMathClamp.h:
```diff
@@ -0,0 +1,63 @@
+    bool equals(const CSSNumericValue&) const final;
+
+    CSSMathClamp(CSSNumericType&&, Ref<CSSNumericValue>&&, Ref<CSSNumericValue>&&, Ref<CSSNumericValue>&&);
+    Ref<CSSNumericValue> m_lower;
```
nit: Would be nice to have a blank line separating data members from functions.

cdumez reviewed (approved):
LGTM with a few comments.

Commit:
Implement most of CSSMathClamp
https://bugs.webkit.org/show_bug.cgi?id=243956

Reviewed by Chris Dumez.

Also update the WPT tests that cover it and the other parts of the css-typed-om.

* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/idlharness-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/perspective-typed-arithmetic-crash.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/resources/testhelper.js:
(assert_style_value_equals):
(createDivWithoutStyle):
(assert_matrix_approx_equals):
(remove_leading_spaces): Deleted.
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/set-css-wide-in-custom-property-crash.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-normalization/normalize-tokens.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-normalization/normalize-tokens.tentative.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-normalization/transformvalue-normalization.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-normalization/transformvalue-normalization.tentative.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-serialization/crashtests/cssInvertValue-convert-crash.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-serialization/crashtests/cssInvertValue-zero.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-serialization/cssMathValue.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-serialization/cssMathValue.tentative.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-serialization/cssTransformValue.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-serialization/cssTransformValue.tentative.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-subclasses/cssColorValue-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-subclasses/cssColorValue.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-subclasses/cssHWB-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-subclasses/cssHWB.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-subclasses/cssPerspective.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-subclasses/cssPerspective.tentative.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-subclasses/numeric-objects/cssMathValue.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-subclasses/numeric-objects/cssMathValue.tentative.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-subclasses/numeric-objects/equals.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-subclasses/numeric-objects/equals.tentative.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-subclasses/numeric-objects/parse.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-subclasses/numeric-objects/parse.tentative.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-subclasses/numeric-objects/to.tentative-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/stylevalue-subclasses/numeric-objects/to.tentative.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/computed/getAll-disconnected-element-crash.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/inline/get-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/inline/get-shorthand-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/inline/get-shorthand.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/inline/get.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/contain.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/container-name-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/container-name.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/container-type-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/container-type.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/font-synthesis.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/overflow-clip-margin-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/overflow-clip-margin.html: Copied from LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/the-stylepropertymap/properties/font-synthesis.html.
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/width-by-clamp-px-em-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-typed-om/width-by-clamp-px-em.html: Added.
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/JSCSSStyleValueCustom.cpp:
(WebCore::toJSNewlyCreated):
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/css/typedom/CSSStyleValue.h:
(WebCore::isCSSNumericValue):
(WebCore::isCSSMathValue):
* Source/WebCore/css/typedom/numeric/CSSMathClamp.cpp: Added.
(WebCore::CSSMathClamp::create):
(WebCore::CSSMathClamp::CSSMathClamp):
(WebCore::CSSMathClamp::serialize const):
(WebCore::CSSMathClamp::toSumValue const):
(WebCore::CSSMathClamp::equals const):
* Source/WebCore/css/typedom/numeric/CSSMathClamp.h: Added.
(isType):
* Source/WebCore/css/typedom/numeric/CSSMathClamp.idl: Copied from Source/WebCore/css/typedom/numeric/CSSMathOperator.idl.
* Source/WebCore/css/typedom/numeric/CSSMathInvert.cpp:
(WebCore::CSSMathInvert::equals const):
* Source/WebCore/css/typedom/numeric/CSSMathOperator.h:
(): Deleted.
* Source/WebCore/css/typedom/numeric/CSSMathOperator.idl:

Canonical link: https://commits.webkit.org/253604@main
